kernel/src/syscall.rs

```diff
@@ -252,53 +255,54 @@ impl Syscall {
     pub fn from_register_arguments(
         syscall_number: u8,
         r0: usize,
-        r1: usize,
-        r2: usize,
-        r3: usize,
+        r1: CapabilityPtr,
+        r2: CapabilityPtr,
+        r3: CapabilityPtr,
     ) -> Option<Syscall> {
         match SyscallClass::try_from(syscall_number) {
             Ok(SyscallClass::Yield) => Some(Syscall::Yield {
                 which: r0,
-                param_a: r1,
-                param_b: r2,
+                param_a: r1.into(),
+                param_b: r2.into(),
             }),
             Ok(SyscallClass::Subscribe) => Some(Syscall::Subscribe {
                 driver_number: r0,
-                subdriver_number: r1,
-                upcall_ptr: r2 as *mut (),
+                subdriver_number: r1.into(),
+                upcall_ptr: r2,
                 appdata: r3,
             }),
             Ok(SyscallClass::Command) => Some(Syscall::Command {
                 driver_number: r0,
-                subdriver_number: r1,
-                arg0: r2,
-                arg1: r3,
+                subdriver_number: r1.into(),
+                arg0: r2.into(),
+                arg1: r3.into(),
             }),
             Ok(SyscallClass::ReadWriteAllow) => Some(Syscall::ReadWriteAllow {
                 driver_number: r0,
-                subdriver_number: r1,
-                allow_address: r2 as *mut u8,
-                allow_size: r3,
+                subdriver_number: r1.into(),
+                allow_address: r2.as_ptr_checked::<u8>(r3.into(), MetaPermissions::ReadWrite)
+                    as *mut u8,
+                allow_size: r3.into(),
             }),
             Ok(SyscallClass::UserspaceReadableAllow) => Some(Syscall::UserspaceReadableAllow {
                 driver_number: r0,
-                subdriver_number: r1,
-                allow_address: r2 as *mut u8,
-                allow_size: r3,
+                subdriver_number: r1.into(),
+                allow_address: r2.as_ptr_checked::<u8>(r3.into(), MetaPermissions::Read) as *mut u8,
+                allow_size: r3.into(),
             }),
             Ok(SyscallClass::ReadOnlyAllow) => Some(Syscall::ReadOnlyAllow {
                 driver_number: r0,
-                subdriver_number: r1,
-                allow_address: r2 as *const u8,
-                allow_size: r3,
+                subdriver_number: r1.into(),
+                allow_address: r2.as_ptr_checked::<u8>(r3.into(), MetaPermissions::Read) as *mut u8,
```

Comment:
I am happy with `as_ptr_checked` returning an `Option`, but I think that opens up the question of how we would like to handle it: by propagating the `None`?

On one hand, that lets the user know about the error.

On the other, the way it exists currently removes the old allow. That means that any commands that a user sends after won't accidentally be made against the previously allowed thing. At least propagating the error means they shouldn't fall victim to this. But its hard to guess at internal logic of userspace and zeroing out the allow felt safe. We could always do `as_ptr_checked().unwrap_or(core::ptr::null())` if we liked this behavior but wanted the option.

Reply:
At a conceptual level, Allow treats buffer pointers as if they are not copyable. When a userspace process invokes Allow, it passes that buffer to the kernel. The kernel should pass that buffer back eventually: either as part of the failure message for this Allow call, or as part of the success message of a future Allow call. You can see this in [TRD 104](https://github.com/tock/tock/blob/f363a97a62b1f58bd81806a2e21af22a39199b45/doc/reference/trd104-syscalls.md#44-read-write-allow-class-id-3):

> The return variants for Read-Write Allow system calls are `Failure with 2 u32` and `Success with 2 u32`.  In both cases, `Argument 0` contains an address and `Argument 1` contains a length. When a driver implementing the Read-Write Allow system call returns a failure result, it MUST return the same address and length as those that were passed  in the call. When a driver implementing the Read-Write Allow system call returns a success result, the returned address and length MUST be those that were passed in the previous call, unless this is the first call. On the first successful invocation of a particular Read-Write Allow system call, an driver implementation MUST return address 0 and size 0.

To be consistent with this design, you need to either return an error or keep the pointer intact so it can be returned on a later Allow. IMO, letting the user know about the error seems like the better option versus pretending to succeed then telling the syscall driver that the user passed a null buffer. It's also consistent with the kernel's behavior if userspace passes a pointer that points outside its address space:

> The Tock kernel MUST check that the passed buffer is contained within the calling process's writeable address space. Every byte of a passed buffer must be readable and writeable by the process. Zero-length buffers may therefore have arbitrary addresses. If the passed buffer is not complete within the calling process's writeable address space, the kernel MUST return a failure result with an error code of `INVALID`.

The userspace libraries should already be compatible with this behavior, and if not, that is a bug in the userspace library.